parquet-hadoop/src/main/java/parquet/hadoop/MemoryManager.java

```diff
@@ -47,6 +49,28 @@
   private final long minMemoryAllocation;
   private final Map<InternalParquetRecordWriter, Long> writerList = new
       HashMap<InternalParquetRecordWriter, Long>();
+  private final Set<CounterCallBack> counterCallBacks = new HashSet<CounterCallBack>();
+
+  /**
+   * A callback interface for upper layer, such as Hive, getting notified when memory usage changes
+   */
+  public static abstract class CounterCallBack {
```

Comment:
I think it would be better to have a generic callback that is used for counters in Hive, rather than having an API here that is focused around counters. Ideally, the Callback would just be a Runnable. Naming should happen when the callback is registered:

```java
final Counter counter = ...;
MemoryManager.registerScaleCallback( "increment-hive-counter", new Runnable() {
    public void run() {
      counter.increment(1);
    }
} );
```